ch5-pwd/main.cpp

```diff
@@ -0,0 +1,115 @@
+/****************************************************
+ * ch5-pwd - Prints the working directory full path *
+ ****************************************************/
+
+#include <cerrno>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <system_error>
+#include <vector>
+
+#include <dirent.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+int main(int /*argc*/, char* argv[]) try
+{
+    auto path = std::vector<std::string>();
+    while (true) // iterate from current WD to root directory
+    {
+        struct stat dotStat;
+        if (::stat(".", &dotStat) < 0)
+        {
+            auto savedErrno = errno;
+            throw std::system_error(savedErrno, std::generic_category(),
+                                    "cannot stat a directory in the path");
+        }
+        struct stat dotDotStat;
+        if (::stat("..", &dotDotStat) < 0)
+        {
+            auto savedErrno = errno;
+            throw std::system_error(savedErrno, std::generic_category(),
+                                    "cannot stat a directory in the path");
+        }
+        if (dotStat.st_ino == dotDotStat.st_ino) // root directory
+        {
+            break;
+        }
+
+        auto dotDotDir = ::opendir("..");
+        if (!dotDotDir)
+        {
+            auto savedErrno = errno;
+            throw std::system_error(savedErrno, std::generic_category(),
+                                    "cannot open a directory in the path");
+        }
+        // iterate over the parent directory to find the name of the current
+        //  directory. if not found, iterate again and stat() each entry, to get
+        //  the 'inner' inode of mount points.
+        for (auto stat = false;;)
```

Comment:
Seems like this loop would be better as a function called twice, once with `stat=false` and again with `stat=true`, to simplify the logic here. The error handling that depends on `stat` value will be done here, to simplify the function logic.

Even better option might be to pass a callable predicate with the content of lines 74-77 or line 84, instead of a boolean argument.

That function might be simplified further by using a generator-style to bring the next `ent`, but this might be too much work without `std::generator` support